---
title: Go 语言：函数
tags:
  - go
categories: go
date: 2023-04-25 20:31:00
---


### 函数

```go
// 声明
func funcName(param1 type1, param2 type2, ...interface{}) (ret1 type1, ret2 type2) {
    // 函数体
}

// 变长参数形参与实参转换
func myAppend(sl []int, elems ...int) []int {
    fmt.Println("%T\n", elems)
    if (len(elems) == 0) {
        fmt.Println("elems is empty")
        return sl
    }

    sl = append(sl, elems...)
    return sl
}

func main() {
    sl := []int{1, 2, 3}
    sl = myAppend(sl)
    fmt.Println(sl)  // [1, 2, 3]  通过切片实现
    sl = myAppend(sl, 4, 5, 6)
    fmt.Println(sl)  // [1, 2, 3, 4, 5, 6] 通过切片实现
}

// 函数返回值
func foo()
func foo() error
func foo() (int, string, error)
```

函数作为“一等公民”的使用。

1. 函数类型的使用
```go
func greeting(w http.ResponseWriter, r *http.Request) {
    fmt.Fprintf(w, "Welcome, Gopher!\n")
}

func main() {
    http.ListenAndServe(":8080", http.HandlerFunc(greeting))
}

// $GOROOT/src/net/http/server.go
type HandlerFunc func(ResponseWriter, *Request)
// ServeHTTP 实现了Handler接口
func (f HandlerFunc) ServeHTTP(w ResponseWriter, r *Request) {
    f(w, r)
}
```
2. 利用闭包简化函数调用
```go
func times1(x, y int) int {
    return x * y
}

// 可以使用闭包简化函数调用
func times2(x int) func(int) int {
    return func(y int) int {
        return x * y
    }
}

func main() {
    timeTwo := times2(2)
    fmt.Println(timeTwo(3))  // 6
    timeThree := times2(3)
    fmt.Println(timeThree(3))  // 9
}
```

### 错误处理

`Go`继承了`C`语言的错误处理方式。
1. 函数返回值
```go

// error 是一个接口类型，通常在函数返回值的末尾位置声明。
func foo() (int, error) {
    return 0, errors.New("error")
}
func main() {
    ret, err := foo()
}
```

在一些场景下，错误处理者需要从错误中提取出更多的信息。这个时候我们可以自定义错误类型来满足这一需求。比如：标准库中`net`包就定义了一种携带额外上下文的错误类型。
```go
// $GOROOT/src/net/net.go
type OpError struct {
    Op string
    Net string
    Source Addr
    Addr Addr
    Err error
}

// $GOROOT/src/net/http/server.go
func isCommonNetReadError(err error) bool {
    if err == io.EOF {
        return true
    }
    if neterr, ok := err.(net.Error); ok && neterr.Timeout() {
        return true
    }
    if oe, ok := err.(*net.OpError); ok && oe.Op == "read" {
        return true
    }

    return false
}

```

`Go`中使用`error`的几点好处：
1. 统一了错误类型
2. 错误是值
3. 易扩展，支持自定错误上下文

错误处理的几种策略:
1. 透明错误处理策略，就是根据函数/方法返回的`error`类型变量中携带的错误信息做相应的处理。
```go
err := doSomething()

if err != nil {
    // 不关心err变量底层错误携带的具体上下文信息
    // 简单处理错误逻辑并返回
    return err
}
```
2. “哨兵”错误处理策略，我们可以通过`errors.Is`方法来判断错误是否为“哨兵”错误。“哨兵”错误值变量以 ErrXXX 格式命名。和透明错误策略相比，“哨兵”策略让错误处理方在有检视错误值的需求时候，可以“有的放矢”。

```go
var ErrSentinel = errors.New("the underlying sentinel error")

func main() {
    err1 := fmt.Errorf("wrap sentinel: %w", ErrSentinel)
    err2 := fmt.Errorf("wrap err1 %w", err1)

    fmt.Println(err2 == ErrSentinel)  // false

    if errors.Is(err2, ErrSentinel) {
        fmt.Println("err2 is ErrSentinel")
        return
    }

    fmt.Println("err2 is not ErrSentinel")
}
```

3. 错误值类型检查策略，我们可以通过`errors.As`方法来判断错误是否为特定类型的错误。
```go
type MyError struct {
    e string
}

func (e *MyError) Error() string {
    return e.e
}

func main() {
    var err = &MyError{"MyError error demo"}
    err1 := fmt.Errorf("wrap err: %w", err)
    err2 := fmt.Errorf("wrap err1: %w", err1)
    var e *MyError
    if errors.As(err2, &e) {
        fmt.Println("MyError is on the chain of err2")
        fmt.Println(e == err)
        return
    }
    fmt.Println("MyError is not on the chain of err2")
}
```

4. 错误行为特征检查策略，将某个包中的错误类型归类，统一提取出一些公共的错误行为特征，并将这些错误行为特征放入一个公开的接口类型中。

```go
type OpError struct {
    // Err is the error that occurred during the operation
    Err error
}

type temporary interface {
    Temporary() bool
}

func (e *OpError) Temporary() bool {
    if ne, ok := e.Err.(*os.SyscallError); ok {
        t, ok := ne.Err.(temporary)
        return ok && t.Temporary()
    }

    t, ok := e.Err.(temporary)
    return ok && t.Temporary()
}
```

函数健壮性设计原则:
1. 不要相信任何外部输入的参数。
- 合法性检查

2. 不要忽略任何一个错误。
- 在调标准度库和第三方函数的时候，要对返回的错误进行处理。

3. 不要假定异常不会发生。

**异常处理**

`Go`语言中使用`panic`和`recover`关键字来实现异常处理。`panic`主要有两类来源：一类来自`Go`运行时，另一类则是`Go`开发人员通过`panic`主动触发的。

```go
// panic 函数的调用
func foo() {
	defer func() {
		if err := recover(); err != nil {  // 捕捉panic
			println("recover in bar")
		}
	}()
    println("call foo")
    bar()
    println("exit foo")
}

func bar() {
    println("call bar")
    panic("panic in bar")
    zoo()
    println("exit bar")
}

func zoo() {
    println("call zoo")
    println("exit zoo")
}

func main() {
    println("call main")
    foo()
    println("exit main")
}
```

**如何应对panic**
1. 评估程序对`panic`的忍受度。
- 如果是命令行程序，我们可以重行运行，执行程序，介是常驻内存的服务，比如：`Web`服务，我们就需要对`panic`处理。
```go
// $GOROOT/src/net/http/server.go
// Serve a new connection.
func (c *conn) serve(ctx context.Context) {
    c.remoteAddr = c.rwc.RemoteAddr().String()
    ctx = context.WithValue(ctx, LocalAddrContextKey, c.rwc.LocalAddr())
    defer func() {
        if err := recover(); err != nil && err != ErrAbortHandler {
            const size = 64 << 10
            buf := make([]byte, size)
            buf = buf[:runtime.Stack(buf, false)]
            c.server.logf("http: panic serving %v: %v\n%s", c.remoteAddr, err, buf)
        }
        if !c.hijacked() {
            c.close()
            c.setState(c.rwc, StateClosed, runHooks)
        }
    }()
    ... ...
}
```
2. 提示潜在的`Bug`。

```go
func (d *decodeState) valueQuoted() interface{} {
    switch d.opcode {
    default:
        panic(phasePanicMsg)

    case scanBeginArray, scanBeginObject:
        d.skip()
        d.scanNext()

    case scanBeginLiteral:
        v := d.literalInterface()
        switch v.(type) {
        case nil, string:
            return v
        }
    }
    return unquotedValue{}
}
```

3. 不能混淆异常与错误
- 在`Java`中，我们常看到有一些定义好的异常类：`IOException`，`EOFException`，这些和`Go`中的哨兵错误类似，都是用来代表特定场景下的错误状态。
- `Go`中，如果是第三方的包，一定不要将`panic`当作错误返回给API调用者。

**`defer`需要注意项**
1. 有些函数可以作为`defer`函数。
```go
append, cap, len, make, new, imag 是不能直接作为`defer`函数的。
clone, copy, delete, print recover 可以直接设置为`defer`函数。
```
2. `defer`关键字后面表达式的求值时机，`defer`函数被调度执行调度时以`LIFO`次序出栈执行。
3. `defer`带来的性能损耗，在1.17版本以上的`Go`版本中，`defer`关键字的性能损耗已经很小了。


